strapi/strapi
# [Custom fields] Create hook to lazy load custom field components

### What does it do

- Removes lazy loading custom field components via `React.lazy` and `React.Suspense` 
- Creates a `useLazyComponents` hook that handles the lazy loading of custom field components 
   - NOTE: This could be modified to handle other components in the future, like the ones registered on the Fields API
- Lazy loads the components at a higher level in the `EditView` page vs the `Inputs` component and returns the `LoadingPageIndicator` until all components have loaded

### Why is it needed

- Custom fields currently cause a bug with InputJSON where the CSS for the json input is not loaded and the formatting doesn’t work properly (mainly the gutters don’t appear)
- `React.lazy` is not intended to be used inside a component, it is designed to be used outside the component with other imports: [https://reactjs.org/docs/code-splitting.html](https://reactjs.org/docs/code-splitting.html), https://beta.reactjs.org/apis/react/lazy
- We were wrapping `React.lazy` in `useMemo` to prevent a re-render. I found similar workarounds [here](https://github.com/facebook/react/issues/14299), but I think this is ill advised since the [useMemo React docs state](https://reactjs.org/docs/hooks-reference.html#usememo):

> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.

- Admittedly I do not fully understand the bug and my fix is based on the assumption that React.lazy + useMemo is not a reliable way to lazy load the components since both are not being used as React intended.
- I found two things that fixed the issue:
    - Moving the `React.Suspense` to be the direct parent of the lazy loaded component (no idea why that worked)
    - Replacing React.lazy / React.Suspense by loading the component asynchronously

### How to test it

- Go to any content-type with a custom field and an a JSON field. Both should work properly. You can also try changing the network speed with dev tools throttling
    - NOTE:  Chrome fast 3g does not fully load the JSON input component with or without a custom field present. This problem does not exist in firefox. I do not believe this is related to custom fields.
- Try installing and using multiple custom fields with the JSON input (you will have to yarn build && yarn develop instead of using the admin on 4000)

### Related issue(s)/PR(s)

fixes https://github.com/strapi/strapi/issues/14489
